File: python/example2.py
from roukf import ROUKF, SigmaDistribution
import numpy as np

def forward_operator(x: np.ndarray, n_states: int, theta: np.ndarray, n_parameters: int) -> int:
    try:
        x[0] = theta[0] * x[0]

        theta[0] = theta[0]
        return 1
    except Exception as e:
        print(f"Error in forward operator: {e}")
        return 0

def observation_operator(x: np.ndarray, n_states: int, z: np.ndarray, n_observations: int) -> None:
    try:
        for i in range(n_observations):
            z[i] = x[i]

    except Exception as e:
        print(f"Error in observation operator: {e}")

def data_generator(numSteps, X0 = np.array([1]), Phi = 0.9):
    timeSeries = np.zeros((1, numSteps))
    x1 = X0[0]
    for i in range(numSteps):
        x1 = Phi * x1
        timeSeries[0, i] = x1

    return timeSeries


num_steps = 10000
observations = data_generator(num_steps)
def main():

    # Example dimensions
    n_observations = 1  # Number of observations
    n_states = 1       # Number of states
    n_parameters = 1    # Number of parameters to estimate

    # Initialize initialGuess and parametersUncertainties
    parameters_uncertainty = np.array([10], dtype=np.float64)

    initial_guess = np.array([1], dtype=np.float64)
    initial_parameters = np.full(n_parameters, 0, dtype=np.float64)


    # Create uncertainty arrays
    states_uncertainty = np.array([1], dtype=np.float64)


    # Initialize ROUKF
    kalman_filter = ROUKF(
        n_observations,
        n_states,
        n_parameters,
        states_uncertainty,
        parameters_uncertainty,
        SigmaDistribution.CANONIC
    )

    initial_state = np.ascontiguousarray(initial_guess)
    initial_parameter = np.ascontiguousarray(initial_parameters)
    print("###########################")
    print(initial_state)
    print("###########################")

    # Set initial condition
    kalman_filter.setState(initial_state)
    kalman_filter.setParameters(initial_parameter)



    for i in range(num_steps):
        # Execute a step of the Kalman filter
        observe = observations[:, i]
        error = kalman_filter.executeStep(
            observe,
            forward_operator, 
            observation_operator
        )
        
        state_estimate = kalman_filter.getState()
        params = kalman_filter.getParameters(n_parameters)


        print(f"Error: {error}, State estimate: {state_estimate[:5]}")  # Print first 5 values
        print(params)



    # Reset the Kalman filter
    kalman_filter.reset(
        n_observations,
        n_states,
        n_parameters,
        states_uncertainty,
        parameters_uncertainty,
        SigmaDistribution.CANONIC
    )
    print("Kalman filter has been reset.")

if __name__ == "__main__":
    main()

File: example2.cpp
#include "ROUKF.h"
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>

int forwardOperator(double* x, int nStates, double* theta, int nParameters);
void observationOperator(double* x, int nStates, double* z, int nObservations);
std::vector<double> dataGenerator(int numSteps, double X0 = 1.0, double Phi = 0.9);

std::vector<double> dataGenerator(int numSteps, double X0, double Phi) {
    std::vector<double> timeSeries(numSteps);
    double x1 = X0;
    for (int i = 0; i < numSteps; i++) {
        x1 = Phi * x1;
        timeSeries[i] = x1;
    }
    return timeSeries;
}

int forwardOperator(double* x, int nStates, double* theta, int nParameters) {
    try {
        x[0] = theta[0] * x[0];
        theta[0] = theta[0];
        return 1;
    } catch (...) {
        std::cerr << "Error in forward operator" << std::endl;
        return 0;
    }
}

void observationOperator(double* x, int nStates, double* z, int nObservations) {
    try {
        for (int i = 0; i < nObservations; i++) {
            z[i] = x[i];
        }
    } catch (...) {
        std::cerr << "Error in observation operator" << std::endl;
    }
}

int main() {
    const int numSteps = 10000;
    const int nObservations = 1;
    const int nStates = 1;
    const int nParameters = 1;
    
    // Generate observations
    std::vector<double> observations = dataGenerator(numSteps);
    
    double parametersUncertainties[nParameters] = {10.0};
    double statesUncertainties[nStates] = {1.0};
    
    double initialGuess[nStates] = {1.0};
    double initialParameters[nParameters] = {0.0};  

    ROUKF* kalmanFilter = new ROUKF(
        nObservations, 
        nStates, 
        nParameters, 
        statesUncertainties,
        parametersUncertainties, 
        SigmaPointsGenerator::SIGMA_DISTRIBUTION::CANONIC
    );

    kalmanFilter->setState(initialGuess);
    kalmanFilter->setParameters(initialParameters);

    // Main loop
    for (int i = 0; i < numSteps; i++) {
        double observe[nObservations] = {observations[i]};
        
        double error = kalmanFilter->executeStep(
            observe,
            forwardOperator, 
            observationOperator
        );
        
        double* stateEstimate = nullptr;
        double* params = nullptr;
        kalmanFilter->getState(&stateEstimate);
        kalmanFilter->getParameters(&params);

        std::cout << "Error: " << error << ", State estimate: " << stateEstimate[0] << std::endl;
        std::cout << "Parameters: " << params[0] << std::endl;
    }

    // Reset the Kalman filter
    kalmanFilter->reset(
        nObservations,
        nStates,
        nParameters,
        statesUncertainties,
        parametersUncertainties,
        SigmaPointsGenerator::SIGMA_DISTRIBUTION::CANONIC
    );
    std::cout << "Kalman filter has been reset." << std::endl;

    // Clean up
    delete kalmanFilter;
    return 0;
}